File: 20. Valid Parentheses/20. Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid2 = function(s) {
    if(s.length%2===1){
        return false;
    }
    var bracket=['{}','[]','()'];
    var f=true;
    while(f){
        f=false;
        for(let i=s.length-2;i>=0;i--){
            if(bracket.includes(s.substr(i,2))){
                s=s.substr(0,i)+s.substring(i+2);
                f=true
            }
        }
    }
    return s.length===0
};

var isValid = function(s) {
    if(s.length%2===1){
        return false;
    }
    var brackets=['{}','[]','()'];
    if(brackets.includes(s)||s===""){
        return true;
    }
    var bracket={
        '{':'}',
        '[':']',
        '(':')',
    };
    let index = s.indexOf(bracket[s[0]]);
    if(index<0){
        return false;
    }
    let lastIndex = s.substring(0,index+1).lastIndexOf(s[0]);
    if(lastIndex<0){
        return false;
    }
    if(!isValid(s.substring(lastIndex+1,index))){
        return false;
    }else{
        return isValid(s.substring(0,lastIndex)+s.substring(index+1))
    }
};

var isValid3 = function(s) {
    if(s.length%2===1){
        return false;
    }
    var stack = [];
    for (var i =0; i < s.length; i++) {
        var popped;
        switch(s[i]) {
            case '{':
                stack.push(s[i]);
                break;
            case '[':
                stack.push(s[i]);
                break;
            case '(':
                stack.push(s[i]);
                break;
            case '}':
                popped = stack.pop();
                if (popped !== '{') {
                    return false;
                }
                break;
            case ']':
                popped = stack.pop();
                if (popped !== '[') {
                    return false;
                }
                break;
            case ')':
                popped = stack.pop();
                if (popped !== '(') {
                    return false;
                }
                break;
        }
    }
    return stack.length==0;
};

var chai = require('chai');
var assert = chai.assert;

describe('isValid2', () => {
  it('test {}[]()', () => {
      assert.isTrue(isValid2('{}[]()'));
  });
  it('test []({}[])[{}()]()([])', () => {
      assert.isTrue(isValid2('[]({}[])[{}()]()([])'));
  });
  it('test {}[(]())', () => {
      assert.isFalse(isValid2('{}[(]())'));
  });
});

describe('isValid', () => {
  it('test {}[]()', () => {
      assert.isTrue(isValid('{}[]()'));
  });
  it('test []({}[])[{}()]()([])', () => {
      assert.isTrue(isValid('[]({}[])[{}()]()([])'));
  });
  it('test {}[(]())', () => {
      assert.isFalse(isValid('{}[(]())'));
  });
});

describe('isValid3', () => {
  it('test {}[]()', () => {
      assert.isTrue(isValid3('{}[]()'));
  });
  it('test []({}[])[{}()]()([])', () => {
      assert.isTrue(isValid3('[]({}[])[{}()]()([])'));
  });
  it('test {}[(]())', () => {
      assert.isFalse(isValid3('{}[(]())'));
  });
});

File: 20. Valid Parentheses/20. Valid Parentheses.java

import java.util.*;


public class Solution {
    public static boolean isValid2(String s) {
        Stack<Integer> p = new Stack<>();
        for(int i = 0; i < s.length(); i++) {
            int q = "(){}[]".indexOf(s.charAt(i));
            if(q % 2 == 1) {
                if(p.isEmpty() || p.pop() != q - 1) return false;
            } else p.push(q);
        }
        return p.isEmpty();
    }
    public static boolean isValid(String s) {
        if (s.isEmpty()) return true;
        char[] chars = s.toCharArray();
        char[] stack = new char[chars.length + 1];
        stack[0] = '-';
        int p = 1;
        for (int i = 0; i < chars.length; i++) {
            char c = chars[i];
            switch(c) {
                case '(':
                    stack[p++] = ')';
                    break;
                case '[':
                    stack[p++] = ']';
                    break;
                case '{':
                    stack[p++] = '}';
                    break;
                default:
                    if (stack[--p] != c) return false;
            }
        }
        return p == 1;
    }
    public static void main(String arg[]) {
        System.out.println(isValid("{}[]()"));
        System.out.println(isValid("[]({}[])[{}()]()([])"));
        System.out.println(isValid("{}[(]())"));
        System.out.println(isValid("[]{}]"));
    }
}
